### PR TITLE
Show diff

### PR DESCRIPTION
**Pull request by Uran198**

### CODE REVIEW

File: coalib/results/Diff.py [hunk 1]
```diff
@@ -100,6 +100,20 @@ def __len__(self):
...
+        Returns total number of changed lines, including lines that will be
+        added as changed.
+        """
+        size = 0
+        for line_diff in self._changes.values():
+            if line_diff.change or line_diff.delete:
+                size += 1
+            if line_diff.add_after:
+                size += len(line_diff.add_after)
+        return size
```
Comment (sils1297) [inline coalib/results/Diff.py]: shouldn't you rather update the `__len__` function to be better? Also you can simplify this: `sum(len(line_diff.add_after for line_diff in self._changes.values() if line_diff.add_after else 1)`

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/e679063679b8e5a81050c153d3fce1a8e7e0356d https://github.com/coala-analyzer/coala/pull/2113/commits/aa6b012eeb8c3e7263a2eaefbe39669317530b73

**Comment by sils1297** [comment]
unack 763c97c

File: coalib/output/ConsoleInteraction.py [hunk 2]
```diff
@@ -87,7 +90,7 @@ def acquire_actions_and_apply(console_printer,
                             key.
     :param cli_actions:     The list of cli actions available.
     """
-    cli_actions = cli_actions or CLI_ACTIONS
+    cli_actions = cli_actions if cli_actions else CLI_ACTIONS
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: uhm why this change? Also unrelated isn't it?

File: coalib/output/ConsoleInteraction.py [hunk 3]
```diff
@@ -225,12 +228,53 @@ def print_result(console_printer,
...
+        if show_patch_action.is_applicable(result, file_dict, file_diff_dict):
+            diff_size = sum(diff.changes_size for diff in result.diffs.values())
+            if diff_size <= DIFF_EXCERPT_MAX_SIZE:
+                if section:
+                    show_patch_action.apply_from_section(result,
+                                                         file_dict,
+                                                         file_diff_dict,
+                                                         section)
+                cli_actions = [x for x in cli_actions
+                               if not isinstance(x, ShowPatchAction)]
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: rather use a filter and a tuple instead of a list IMO

File: coalib/output/ConsoleInteraction.py [hunk 4]
```diff
@@ -37,6 +39,7 @@
                PrintDebugMessageAction(),
                PrintMoreInfoAction(),
                ShowPatchAction()]
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: I think this should be converted to a tuple for immutability, also allows you to get rid of the deepcopy below

File: coalib/output/ConsoleInteraction.py [hunk 5]
```diff
@@ -225,12 +228,53 @@ def print_result(console_printer,
     console_printer.print(format_lines(result.message), delimiter="\n")
 
     if interactive:
+        cli_actions = copy.deepcopy(CLI_ACTIONS)
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: why not a flat copy?

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2113/commits/9534e829028f0703e160f757f0c9af764fe84882 but it looks promising, I'd love to get an asciinema to see it in action!

File: coalib/results/Diff.py (hunk 1 above)
Comment (Uran198) [inline coalib/results/Diff.py]: The `__len__` may be used already in many places with the current intention and tests may not cover it, so changing it can be risky. And to rewrite with sum it'll be
```python
sum(len(line_diff.add_after) for line_diff in self._changes.values()
        if line_diff.add_after else 1) + \
len(line_diff for line_diff in self._changes.values()
        if line_diff.change or line_diff.delete)
```
I'm not sure this is simpler to read.

File: coalib/output/ConsoleInteraction.py (hunk 2 above)
Comment (Uran198) [inline coalib/output/ConsoleInteraction.py]: More explicit IMO, maybe separate commit?

Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: IMO just not do it, if ou want it put it in an extra commit

File: coalib/results/Diff.py (hunk 1 above)
Comment (sils1297) [inline coalib/results/Diff.py]: argh yeah keep it. I think you should at least grep a bit for len and see how it's used.

Comment (Uran198) [inline coalib/results/Diff.py]: Ok, len seems to be used only in tests, so I'll change that.

File: coalib/output/ConsoleInteraction.py [hunk 6]
```diff
@@ -225,12 +227,53 @@ def print_result(console_printer,
...
+                print_diffs_info(result.diffs, file_dict, file_diff_dict,
+                                 console_printer)
         acquire_actions_and_apply(console_printer,
                                   log_printer,
                                   section,
                                   file_diff_dict,
                                   result,
-                                  file_dict)
+                                  file_dict,
+                                  cli_actions)
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: isn't cli_actions undefined `if not interactive`?

File: coalib/output/ConsoleInteraction.py [hunk 7]
```diff
@@ -225,12 +227,53 @@ def print_result(console_printer,
     console_printer.print(format_lines(result.message), delimiter="\n")
 
     if interactive:
+        cli_actions = CLI_ACTIONS
+        show_patch_action = ShowPatchAction()
+        if show_patch_action.is_applicable(result, file_dict, file_diff_dict):
+            diff_size = sum(len(diff) for diff in result.diffs.values())
+            if diff_size <= DIFF_EXCERPT_MAX_SIZE:
+                if section:
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: how can `not section` be?

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/07781982a40075bf5bb066b19fb5a71306a00c04 https://github.com/coala-analyzer/coala/pull/2113/commits/eba0b25381f719fe3a373b78e8522401d7dbc863

File: coalib/output/ConsoleInteraction.py [hunk 8]
```diff
@@ -225,12 +227,53 @@ def print_result(console_printer,
...
+            current_file = original_file
+            new_file = diff.modified
+        unified_diff = difflib.unified_diff(current_file, new_file)
+        additions = 0
+        deletitions = 0
+        for line in unified_diff:
+            if line.startswith('+') and not line.startswith('+++'):
+                additions += 1
+            elif line.startswith('-') and not line.startswith('---'):
+                deletitions += 1
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: I think `additions = sum(1 for line in unified_diff if line.startswith('+ '))` is way better readable

Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: ahm there is no space, huh? The `+++` might be wrong also since I might add a line `++ somethign` and then you wouldn't count that, mabe you can tell difflib to not output the `+++` `---` lines *or* just remove the first 3ish lines from the unified diff?

File: coalib/output/ConsoleInteraction.py [hunk 9]
```diff
@@ -225,12 +227,53 @@ def print_result(console_printer,
...
+
+def print_diffs_info(diffs, original_file_dict, file_diff_dict, printer):
+    for filename, diff in sorted(diffs.items()):
+        original_file = original_file_dict[filename]
+        if filename in file_diff_dict:
+            current_file = file_diff_dict[filename].modified
+            new_file = (file_diff_dict[filename] + diff).modified
+        else:
+            current_file = original_file
+            new_file = diff.modified
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: I'm pretty sure this is duplication and should go into a function in the ShowPatchAction or so?

File: coalib/output/ConsoleInteraction.py (hunk 6 above)
Comment (Uran198) [inline coalib/output/ConsoleInteraction.py]: Yes, and this function wouldn't be called, so no problem

File: coalib/output/ConsoleInteraction.py (hunk 7 above)
Comment (Uran198) [inline coalib/output/ConsoleInteraction.py]: Not sure, what you mean here. If it's passed as `None` to this function.

File: coalib/output/ConsoleInteraction.py (hunk 6 above)
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: ah right, sorry

File: coalib/output/ConsoleInteraction.py (hunk 9 above)
Comment (Uran198) [inline coalib/output/ConsoleInteraction.py]: It is. But it will be redundant to make it a function, though I'll change this a little.

Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: if a function is redundant it's not duplication, right?

File: coalib/output/ConsoleInteraction.py [hunk 10]
```diff
@@ -218,19 +221,54 @@ def print_result(console_printer,
...
+
+
+def print_diffs_info(diffs, original_file_dict, file_diff_dict, printer):
+    for filename, diff in sorted(diffs.items()):
+        if filename in file_diff_dict:
+            current_file = file_diff_dict[filename].modified
+        else:
+            current_file = original_file_dict[filename]
+        new_file = (diff + file_diff_dict.get(filename, Diff([]))).modified
+        unified_diff = list(difflib.unified_diff(current_file, new_file))[2:]
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: why not tuple?

File: coalib/output/ConsoleInteraction.py [hunk 11]
```diff
@@ -218,19 +221,54 @@ def print_result(console_printer,
...
+                                  cli_actions)
+
+
+def print_diffs_info(diffs, original_file_dict, file_diff_dict, printer):
+    for filename, diff in sorted(diffs.items()):
+        if filename in file_diff_dict:
+            current_file = file_diff_dict[filename].modified
+        else:
+            current_file = original_file_dict[filename]
+        new_file = (diff + file_diff_dict.get(filename, Diff([]))).modified
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: I'd rather put this in the if/else to avoid the unneeded creation of a diff; `Diff([])` will be executed

File: coalib/output/ConsoleInteraction.py [hunk 12]
```diff
@@ -218,19 +221,54 @@ def print_result(console_printer,
...
+        unified_diff = list(difflib.unified_diff(current_file, new_file))[2:]
+        additions = sum(1 for line in unified_diff if line.startswith('+'))
+        deletitions = sum(1 for line in unified_diff if line.startswith('-'))
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: typo, deletitions

File: coalib/output/ConsoleInteraction.py (hunk 10 above)
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: actually, rather than creating a unified diff and all that it'd be way better and more useful to add a diffstat function to the Diff class and just sum up the added lines, deleted lines and for a change add 1 to both values. No need to fiddle around with file dicts :)

File: tests/output/ConsoleInteractionTest.py [hunk 13]
```diff
@@ -183,6 +186,68 @@ def test_require_settings(self):
...
+        previous_diffs = {"a": Diff(file_dict['a'])}
+        previous_diffs["a"].change_line(2, "b\n", "b_changed\n")
+        with retrieve_stdout() as stdout:
+            print_diffs_info(diff_dict,
+                             file_dict,
+                             previous_diffs,
+                             self.console_printer),
+            self.assertEqual(stdout.getvalue(),
+                             "|    | +1 -1 in a\n"
+                             "|    | +1 -0 in b\n")
```
Comment (sils1297) [inline tests/output/ConsoleInteractionTest.py]: isn't this case almost the same as the one in 229? Merge them?

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/93e833924ea410eb627f947dde1121e2f094d194

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2113/commits/459000e955485a38a41272dc0748fe65e0204837

File: tests/output/ConsoleInteractionTest.py (hunk 13 above)
Comment (Uran198) [inline tests/output/ConsoleInteractionTest.py]: I believe that tests should be as separate as possible and it's not big deal if there is repetitions in them.

Comment (sils1297) [inline tests/output/ConsoleInteractionTest.py]: ok

Sincerely,

Lasse Schuirmann

contact@viperdev.io
http://viperdev.io/

2016-05-06 16:10 GMT+02:00 Attila <notifications@github.com>:

File: coalib/output/ConsoleInteraction.py [hunk 14]
```diff
@@ -218,19 +219,47 @@ def print_result(console_printer,
...
                                   log_printer,
                                   section,
                                   file_diff_dict,
                                   result,
-                                  file_dict)
+                                  file_dict,
+                                  cli_actions)
+
+
+def print_diffs_info(diffs, original_file_dict, file_diff_dict, printer):
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: unused params, please remove

File: coalib/output/ConsoleInteraction.py [hunk 15]
```diff
@@ -218,19 +219,47 @@ def print_result(console_printer,
                          "not a valid derivative of the coala result "
                          "class.")
         return
+    if section is None:
+        raise ValueError("Section has to be intializied.")
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: if you want this, this has to be added in an own commit and please use enforce_signature for this one, it shouldn't ever happen and thus is a programmer - not runtime - error

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/98ffa8ecde577534dbda41f92d1900626d90dd17 https://github.com/coala-analyzer/coala/pull/2113/commits/9fd2b1b506ba0da09e1ba901d44693c419495bdd https://github.com/coala-analyzer/coala/pull/2113/commits/a830422a487b1047b7951c6b4f716cb70a59e4d3

**Comment by sils1297** [comment]
@Uran198 I'd remove https://github.com/coala-analyzer/coala/pull/2113/commits/e52530afbc8f9c908b9e5a4141148195eaf7732d6 completely, you can add enforce signature but that doesn't need a test
essentially enforce signature is a test
@Uran198 just remove the params and the section check and we can merge it

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/a830422a487b1047b7951c6b4f716cb70a59e4d3

File: tests/output/ConsoleInteractionTest.py [hunk 16]
```diff
@@ -208,16 +269,8 @@ def test_print_result(self):
...
-                    lambda *args: True)
-                print_result(self.console_printer,
-                             self.log_printer,
-                             None,
-                             self.file_diff_dict,
-                             Result("origin", "msg", diffs={
-                                    testfile_path: diff}),
-                             file_dict)
+            ApplyPatchAction.is_applicable = staticmethod(
+                lambda *args: True)
```
Comment (sils1297) [inline tests/output/ConsoleInteractionTest.py]: why are you changing this?

File: coalib/output/ConsoleInteraction.py [hunk 17]
```diff
@@ -497,9 +497,6 @@ def get_action_info(section, action, failed_actions):
     """
     params = action.non_optional_params
 
-    if section is None:
-        raise ValueError("section has to be intializied.")
-
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: this commit https://github.com/coala-analyzer/coala/pull/2113/commits/803b2af2d0a99e22f81bd0b9eea41eb1a620736f doesn't contain test changes despite commit message claiming that

**Comment by sils1297** [comment]
(cancelled your build because something else is to be merged)

File: tests/output/ConsoleInteractionTest.py (hunk 16 above)
Comment (Uran198) [inline tests/output/ConsoleInteractionTest.py]: This basically tested that get_action_info will raise ValueError when passed section == None. Maybe, it's better for this to go into https://github.com/coala-analyzer/coala/pull/2113/commits/803b2af2d0a99e22f81bd0b9eea41eb1a620736f

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/cc2a2008734d6bc048b97b5e1ec05f2234467d2c

**Comment by sils1297** [comment]
reack https://github.com/coala-analyzer/coala/pull/2113/commits/354e362d018297ac2ac58882ac9a34b12de507f8

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/3be2948047f5edf7567469fd4f190e5c3399e151

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/2c62d723bf1010bfd8a8e8695a03f599e015c1a6

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2113/commits/a20a8673f8d0d4092a48f1f9f18ef49dbaf3e99d

**Comment by sils1297** [comment]
@rultor merge

**Comment by sils1297** [comment]
@Uran198 can you make an asciinema demoing this? It's awesome. :)

**Comment by sils1297** [comment]
or maybe @sims1253 ? https://github.com/coala-analyzer/coala/pull/2113#issuecomment-217613561
